From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Taiyou06 <kaandindar21@gmail.com>
Date: Fri, 8 Nov 2024 00:06:34 +0100
Subject: [PATCH] Lithium: cached iterate outwards

By: 2No2Name <2No2Name@web.de>
As part of: Lithium (https://github.com/CaffeineMC/lithium)
Licensed under: LGPL-3.0-only (https://www.gnu.org/licenses/lgpl-3.0.html)

Co-authored-by: MrlingXD <90316914+wling-art@users.noreply.github.com>

diff --git a/net/minecraft/core/BlockPos.java b/net/minecraft/core/BlockPos.java
index c183047c17386d78f04120b390abcc6f542c7615..8d176d34218d453eb44632e3ce9ef16733b2c271 100644
--- a/net/minecraft/core/BlockPos.java
+++ b/net/minecraft/core/BlockPos.java
@@ -320,6 +320,11 @@ public class BlockPos extends Vec3i {
         };
     }
 
+    // JettPack start - lithium: cached iterate outwards
+    private static final org.dreeam.leaf.util.cache.IterateOutwardsCache ITERATE_OUTWARDS_CACHE = new org.dreeam.leaf.util.cache.IterateOutwardsCache(50);
+    private static final it.unimi.dsi.fastutil.longs.LongList HOGLIN_PIGLIN_CACHE = ITERATE_OUTWARDS_CACHE.getOrCompute(8, 4, 8);
+    // JettPack end - lithium: cached iterate outwards
+
     public static Iterable<BlockPos> withinManhattan(final BlockPos origin, final int reach) {
         return manhattanOrdered(origin, reach, reach, reach, reach);
     }
@@ -329,6 +334,12 @@ public class BlockPos extends Vec3i {
     }
 
     public static Iterable<BlockPos> withinBoxByManhattanDistance(final BlockPos origin, final int reachX, final int reachY, final int reachZ) {
+        // JettPack start - lithium: cached iterate outwards
+        if (origin != org.dreeam.leaf.util.cache.IterateOutwardsCache.POS_ZERO) {
+            final it.unimi.dsi.fastutil.longs.LongList positions = reachX == 8 && reachY == 4 && reachZ == 8 ? HOGLIN_PIGLIN_CACHE : ITERATE_OUTWARDS_CACHE.getOrCompute(reachX, reachY, reachZ);
+            return new org.dreeam.leaf.util.cache.LongList2BlockPosMutableIterable(origin, positions);
+        }
+        // JettPack end - lithium: cached iterate outwards
         return manhattanOrdered(origin, reachX, reachY, reachZ, reachX + reachY + reachZ);
     }
 
diff --git a/net/minecraft/world/level/levelgen/feature/DeltaFeature.java b/net/minecraft/world/level/levelgen/feature/DeltaFeature.java
index a3b0cd69a2f724ecd99da3e9564fe27cb5b92416..efbd00c7918ba007ce3ebdae8beeca5999c52600 100644
--- a/net/minecraft/world/level/levelgen/feature/DeltaFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/DeltaFeature.java
@@ -47,7 +47,7 @@ public record DeltaFeature(BlockState contents, BlockState rim, IntProvider size
         int radiusZ = this.size.sample(random);
         int radiusLimit = Math.max(radiusX, radiusZ);
 
-        for (BlockPos pos : BlockPos.withinBoxByManhattanDistance(origin, radiusX, 0, radiusZ)) {
+        for (BlockPos pos : BlockPos.withinClippedManhattan(origin, radiusX, 0, radiusZ)) { // Leaf - Lithium - cached iterate outwards
             if (pos.distManhattan(origin) > radiusLimit) {
                 break;
             }
diff --git a/net/minecraft/world/level/levelgen/feature/ReplaceBlobsFeature.java b/net/minecraft/world/level/levelgen/feature/ReplaceBlobsFeature.java
index 270065e06e9b49faa3637aa18b6690ae8a2c57a8..4cd03ba1915460e5aca305cbd838765ea8d31602 100644
--- a/net/minecraft/world/level/levelgen/feature/ReplaceBlobsFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/ReplaceBlobsFeature.java
@@ -43,7 +43,7 @@ public record ReplaceBlobsFeature(BlockState targetState, BlockState replaceStat
         int maximumRadius = Math.max(radiusX, Math.max(radiusY, radiusZ));
         boolean replacedAny = false;
 
-        for (BlockPos pos : BlockPos.withinBoxByManhattanDistance(centerPos, radiusX, radiusY, radiusZ)) {
+        for (BlockPos pos : BlockPos.withinClippedManhattan(centerPos, radiusX, radiusY, radiusZ)) { // Leaf - Lithium - cached iterate outwards
             if (pos.distManhattan(centerPos) > maximumRadius) {
                 break;
             }
